nixos/modules/services/hardware/upower.nix

```diff
@@ -179,24 +179,59 @@ in
         '';
       };
 
+      allowRiskyCriticalPowerAction = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Enable the risky critical power actions "Suspend" and "Ignore".
+        '';
+      };
+
       criticalPowerAction = lib.mkOption {
-        type = lib.types.enum [ "PowerOff" "Hibernate" "HybridSleep" ];
+        type = lib.types.enum [
+          "PowerOff"
+          "Hibernate"
+          "HybridSleep"
+          "Suspend"
+          "Ignore"
+        ];
         default = "HybridSleep";
         description = ''
           The action to take when `timeAction` or
           `percentageAction` has been reached for the batteries
-          (UPS or laptop batteries) supplying the computer
+          (UPS or laptop batteries) supplying the computer.
+
+          When set to `Suspend` or `Ignore`, `allowRiskyCriticalPowerAction`
+          must be set to `true`.
         '';
       };
 
     };
 
   };
 
-
   ###### implementation
 
   config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion =
+          with builtins;
+          let
+            riskyActions = [
+              "Suspend"
+              "Ignore"
+            ];
+            riskyActionEnabled = elem cfg.criticalPowerAction riskyActions;
+          in
+          !cfg.allowRiskyCriticalPowerAction && riskyActionEnabled;
```

Comment:
nit, feel free to ignore: would this be cleaner written as `riskyActionEnabled -> cfg.allowRiskyCriticalPowerAction`? I think it's the same boolean logic, and reads a bit more nicely ("if a risky action is enabled, then we must allow risky actions").

Reply:
Actually, is the current assertion correct? I'm trying out this patch locally, and I've set `services.upower.criticalPowerAction = "Suspend"`, so `riskyActionEnabled = true`. I have not yet set `services.upower.allowRiskyCriticalPowerAction` to anything, so it's just the default value of `false`.

Plugging that into this expression, I get:

`!cfg.allowRiskyCriticalPowerAction && riskyActionEnabled` => `!false && true` => `true && true` => `true`

That's wrong, right? Don't we want this configuration to trigger the assertion (i.e., evaluate to `false`)?

I think the `riskyActionEnabled -> cfg.allowRiskyCriticalPowerAction` I suggested above is correct. Or perhaps `!riskyActionEnabled or cfg.allowRiskyCriticalPowerAction`.

Reply:
TIL `->`.

Quite right. I was treating that as an expression that when evaluated as `true` would trigger an error, rather than (what should have been clear to me) when the assertion condition is not met.

Good spot!